File: gradle-plugins/elsa/src/main/java/com/gridnine/elsa/gradle/plugin/ElsaTypesExtension.java
/*****************************************************************
 * Gridnine http://www.gridnine.com
 * Project: Elsa
 *****************************************************************/

package com.gridnine.elsa.gradle.plugin;

import com.gridnine.elsa.gradle.config.ElsaTypesExtensionData;
import com.gridnine.elsa.gradle.config.ElsaTypesFileData;
import com.gridnine.elsa.gradle.config.ElsaTypesProjectData;
import org.gradle.api.Project;

public class ElsaTypesExtension {

    private final Project project;

    private final ElsaTypesExtensionData data;

    public ElsaTypesExtension(Project project, ElsaTypesExtensionData data ) {
        this.project = project;
        this.data = data;
    }

    private ElsaTypesProjectData findItem() {
        var item = data.items.stream().filter(it -> it.project.equals(project)).findFirst().orElse(null);
        if(item == null){
            item = new ElsaTypesProjectData();
            item.project = project;
            data.items.add(item);
        }
        return item;

    }
    public void destDir(String dir){
        findItem().destDir = project.file(dir);

    }
    public void serialization(String path, String configuratorClassName){
        findItem().serializableTypes.add(new ElsaTypesFileData(project.file(path), configuratorClassName));
    }

    public void domain(String path, String configuratorClassName){
        findItem().domainTypes.add(new ElsaTypesFileData(project.file(path),configuratorClassName));
    }

    public void xsdsLocation(String path){
        data.xsdsLocation = project.file(path);
    }

    public ElsaTypesExtensionData getData() {
        return data;
    }
}

File: gradle-plugins/elsa/src/main/kotlin/com/gridnine/elsa/gradle/plugin/ElsaTypesExtension.kt
package com.gridnine.elsa.gradle.plugin

import com.gridnine.elsa.gradle.config.ElsaTypesExtensionData
import com.gridnine.elsa.gradle.config.ElsaTypesFileData
import com.gridnine.elsa.gradle.config.ElsaTypesProjectData
import org.gradle.api.Project

@ElsaJavaConfigMarker
open class ElsaTypesExtension(private val project: Project, val data: ElsaTypesExtensionData) {

    private fun findItem(): ElsaTypesProjectData {
        var item = data.items.stream().filter { it: ElsaTypesProjectData -> it.project == project }
            .findFirst().orElse(null)
        if (item == null) {
            item = ElsaTypesProjectData()
            item.project = project
            data.items.add(item)
        }
        return item
    }

    fun destDir(dir: String?) {
        findItem().destDir = project.file(dir)
    }

    fun serialization(path: String?, configuratorClassName: String?) {
        findItem().serializableTypes.add(ElsaTypesFileData(project.file(path), configuratorClassName))
    }

    fun domain(path: String?, configuratorClassName: String?) {
        findItem().domainTypes.add(ElsaTypesFileData(project.file(path), configuratorClassName))
    }

    fun custom(path: String?, configuratorClassName: String?) {
        findItem().customTypes.add(ElsaTypesFileData(project.file(path), configuratorClassName))
    }

    fun l10n(path: String?, configuratorClassName: String?) {
        findItem().l10nTypes.add(ElsaTypesFileData(project.file(path), configuratorClassName))
    }

    fun xsdsLocation(path: String?) {
        data.xsdsLocation = project.file(path)
    }

    fun xsdsCustomizationSuffix(suffix: String?) {
        data.xsdsCustomizationSuffix = suffix
    }
}
